README.txt

This is an implementation of the lojban gismu creation algorithm as
described in "The Complete Lojban Language", chapter 4, section 14.

The current version is based on work by Arnt Richard Johansen, integrating
a pure python implementation of the LCS (longest common subsequence)
algorithm for easy installation and optimal performance.

USAGE
=====

To generate scored gismu candidates, run the "gismu_score.py" program,
passing as arguments phonetically lojbanized words.

  python gismu_score.py uan rakan ekspekt esper predpologa mulud

The top ten highest scoring candidates will be displayed. Candidates
are scored according to resemblance to the input words, with additional
consideration given to languages which are more widely spoken in the
world. The default languages and and weighting factors, as derived from
the 1995 Encyclopedia Brittanica Book of the Year, are:

    Chinese     0.347
    Hindi       0.196
    English     0.160
    Spanish     0.123
    Russian     0.089
    Arabic      0.085

OPTIONS
=======

You may pass options to gismu_score.py to modify the way that gismu candidates
are generated and scored.

The --all-letters (-a) option generates candidates using all letters in the
lojban alphabet, whether or not they appear in the input words. This has the
effect of requiring substantially more processing time to generate scores.

The --weights (-w) option controls the language weights, accepting a
comma-separated list of weights, or alternately, a year for which lojban
language weights were published. Years currently supported: 1985, 1987,
1994, 1995 (default), and 1999.

The --shapes (-s) option enables you to experiment with different gismu
"shapes". Pass a comma-separated list of shapes, described with "c"
for consonant and "v" for vowel. The default value for this option is
"ccvcv,cvccv".

The --number-workers (-n) option controls the number of python scoring threads
to use. This may only be useful with python implementations which don't use
a GIL (Global Interpreter Lock) such as jython.

The --deduplicate (-d) option accepts a path to file containing a list of
pre-existing gismu, one per line (e.g "gismu-list.txt"). Candidates will be
matched against the gismu in this file; candidates that are deemed similar
to existing gismu will be disqualified.

The --output (-o) option accepts a filepath. All candidates, along with their
scores will be written to this path. The format is serialized ("marshaled")
tuples, and may be passed as input to "gismu_best.py", e.g.:

  python gismu_best.py < scores.data

The --quiet (-q) option suppresses the display of progress while scores
are being calculated.

CHANGES
=======

See "CHANGELOG.txt".

LICENSE
=======

The scripts and modules in this implementation may be copied, modified,
and distributed under the terms of the GNU General Public License v3.
For details, see "LICENSE.txt".

"gismu-list.txt" contains public domain content furnished by:

  Logical Language Group, Inc.
  2904 Beau Lane
  Fairfax, VA 22031
  USA



CHANGELOG.txt
Version 0.5

Added candidate evaluation to "gismu_score.py".

When generate candidates with custom shapes, excluding forbidden
consonant triples. Also excluding internal sequences which could be
mistaken for "ccvcv" gismu.

Deduplication and serialization of candidates are now options which are
disabled by default.

----

Version 0.4

Replaced mlpy implementation of LCS (longest common subsequence)
algorithm with a pure-python implementation to simplify
installation and enable compatibility with python implementations
such as pypy and jython.

By default, only generating candidates from letters which appear in input
words. This substantially reduces processing time by avoiding the calculation
of scores for less-relevant candidates.

Added command-line options to "gismu_score.py" to customize language weights,
gismu candidate shape, gismu alphabet, and number of scoring worker threads.

Output of "gismu_score.py" / input of "gismu_best.py" now uses python
"marshal" format due to OutOfMemory errors with pickle/cPickle under
jython.

The gismu candidates DBM database used by previous versions of the script
("candidates.db") is no longer created or consulted.

----

Version 0.3

Corrected algorithm to factor in the length of input words per CLL: "The scores
were divided by the length of the source-language word in its Lojbanized form"

----

Version 0.2

Integrated mlpy implementation of LCS (longest common subsequence) algorithm
for better performance.

----

Version 0.1

Arnt Richard Johansen's original implementation as downloaded from:

  http://www.lojban.org/tiki/gismu+scoring+algorithm



gismu_best.py
#!/usr/bin/env python

# Lojban gismu candidate score evaluation script
# Version 0.5

# Copyright 2014 Riley Martinez-Lynch, except where
# Copyright 2012 Arnt Richard Johansen.
# Distributed under the terms of the GPL v3.

# Usage:
#
#   python gismu_score.py -o scores.data uan rakan ekspekt esper predpologa mulud
#   python gismu_best.py < scores.data
#

import platform
import sys

from marshal import load

from gismu_utils import GismuMatcher

def main(scores, gismus):

    print >>sys.stderr, "Sorting scores..."
    scores.sort(lambda x,y:cmp(y[0], x[0]))

    print >>sys.stderr, ""
    print >>sys.stderr, "10 first gismu candidates are:"
    print >>sys.stderr, ""
    for record in scores[:10]:
        print >>sys.stderr, record

    print >>sys.stderr, ""
    print >>sys.stderr, "Excluding candidates similar to existing gismu..."
    matcher = GismuMatcher(gismus)
    for (score, candidate, _) in scores:
        gismu = matcher.find_similar_gismu(candidate)
        if gismu == None:
            print >>sys.stderr, "The winner is....\n"
            print candidate.upper()
            print >>sys.stderr, ""
            break
        else:
            print >>sys.stderr, \
              "Candidate '%s' too much like gismu '%s'." % (candidate, gismu)
    else:
        print >>sys.stderr, "No suitable candidates in top 10 scores."

if __name__ == '__main__':

    gismu_path = 'gismu-list.txt'
    print >>sys.stderr, "Reading list of gismu... "
    gismus = [line.strip() for line in file(gismu_path)]

    print >>sys.stderr, "Loading scores... ",
    scores = load(sys.stdin)
    print >>sys.stderr, "%d scores loaded." % len(scores)

    main(scores, gismus)



gismu_utils.py

# Lojban gismu candidate generation and scoring utilities
# Version 0.5

# Copyright 2014 Riley Martinez-Lynch, except where
# Copyright 2012 Arnt Richard Johansen.
# Distributed under the terms of the GPL v3.

import sys
import re

from itertools import chain, ifilter

# placate python 2.5 (e.g. jython)
if sys.version_info < (2, 6):
    def next(iterator, default):
        try:
            return iterator.next()
        except StopIteration:
            return default

C = 'bcdfgjklmnprstvxz'
V = 'aeiou'

CCVCV = [ C, C, V, C, V ]
CVCCV = [ C, V, C, C, V ]

VALID_CC_INITIALS = [
  'bl', 'br',
  'cf', 'ck', 'cl', 'cm', 'cn', 'cp', 'cr', 'ct',
  'dj', 'dr', 'dz',
  'fl', 'fr',
  'gl', 'gr',
  'jb', 'jd', 'jg', 'jm', 'jv',
  'kl', 'kr',
  'ml', 'mr',
  'pl', 'pr',
  'sf', 'sk', 'sl', 'sm', 'sn', 'sp', 'sr', 'st',
  'tc', 'tr', 'ts',
  'vl', 'vr',
  'xl', 'xr',
  'zb', 'zd', 'zg',
  'zm', 'zv'
]

FORBIDDEN_CC  = [ 'cx', 'kx', 'xc', 'xk', 'mz' ]

FORBIDDEN_CCC = [ 'ndj', 'ndz', 'ntc', 'nts' ]

SIBILANT = 'cjsz'
VOICED   = 'bdgvz'
UNVOICED = 'cfkpstx'

SIMILARITIES = {
  'b':'pv',
  'c':'js',
  'd':'t',
  'f':'pv',
  'g':'kx',
  'j':'cz',
  'k':'gx',
  'l':'r',
  'm':'n',
  'n':'m',
  'p':'bf',
  'r':'l',
  's':'cz',
  't':'d',
  'v':'bf',
  'x':'gk',
  'z':'js'
}

LANGUAGE_WEIGHTS = {

  # Order: Chinese, Hindi, English, Spanish, Russian, Arabic

  '1985'     : (0.360, 0.160, 0.210, 0.110, 0.090, 0.070),
  # http://dag.github.io/cll/4/14

  '1987'     : (0.360, 0.156, 0.208, 0.116, 0.087, 0.073),
  '1994'     : (0.348, 0.194, 0.163, 0.123, 0.088, 0.084),
  # http://www.lojban.org/files/etymology/langstat.94

  '1995'     : (0.347, 0.196, 0.160, 0.123, 0.089, 0.085), # default
  # http://www.lojban.org/files/etymology/langstat.95

  '1999'     : (0.334, 0.195, 0.187, 0.116, 0.081, 0.088)
  # http://www.lojban.org/files/etymology/langstat.99

}

XADD = lambda a,b: [x+y for x in a for y in b]

# LCS implementation adapted from python dynamic programming example at:
#   http://rosettacode.org/wiki/Longest_common_subsequence

def lcs_length(a, b):
    return lcs_matrix(a, b)[-1][-1]

def lcs_matrix(a, b):
    matrix = [[0 for j in range(len(b)+1)] for i in range(len(a)+1)]
    for i, x in enumerate(a):
        for j, y in enumerate(b):
            if x == y:
                matrix[i+1][j+1] = matrix[i][j] + 1
            else:
                matrix[i+1][j+1] = max(matrix[i+1][j], matrix[i][j+1])
    return matrix

class GismuGenerator:

    def __init__(self, c, v, shape_strings):
        self.c = c
        self.v = v
        self.shape_strings = shape_strings

    def iterator(self):
        iterators = [ self.shape_iterator(str) for str in self.shape_strings ]
        return chain(*iterators)

    def shape_iterator(self, shape_string):
        shape = self.shape_for_string(shape_string)
        validator = self.shape_validator(shape_string)
        return ifilter(validator, reduce(XADD, shape))

    def shape_for_string(self, string):
        shape = []
        for letter in list(string.lower()):
          if letter == 'c':
              shape.append(self.c)
          elif letter == 'v':
              shape.append(self.v)
        return shape

    def shape_validator(self, shape):
        predicates = []
        slen = len(shape)
        for i, c in enumerate(shape[:-1].lower()):
            if (c == 'c'):
                if (shape[i+1] == 'c'):
                    predicates.append(self.validator_for_cc(i))
                    if (i < (slen-2)) and (shape[i+2] == 'c'):
                        predicates.append(self.validator_for_ccc(i))
                    if (i < (slen-4)) and (i>0) and (shape[i:i+5] == 'ccvcv'):
                        predicates.append(self.invalidator_for_initial_cc(i))
        return self.validator_for_predicates(predicates)

    def validator_for_cc(_, i):
        if i == 0:
            return lambda x: x[:2] in VALID_CC_INITIALS
        else:
            j = i + 1
            return lambda x: x[i] != x[j] and \
              not (x[i] in VOICED   and x[j] in UNVOICED) and \
              not (x[i] in UNVOICED and x[j] in VOICED  ) and \
              not (x[i] in SIBILANT and x[j] in SIBILANT) and \
              x[i:j + 1] not in FORBIDDEN_CC

    def validator_for_ccc(_, i):
        return lambda x: x[i:i+3] not in FORBIDDEN_CCC

    def invalidator_for_initial_cc(_, i):
        return lambda x: x[i:i+2] not in VALID_CC_INITIALS

    def validator_for_predicates(_, predicates):
        if len(predicates) == 0:
            return lambda x: True
        elif len(predicates) == 1:
            return predicates[0]
        else:
            return lambda x: \
              not next((True for p in predicates if p(x) == False), False)

class GismuScorer:

    def __init__(self, input_words, weights):
        self.input_words_chars = [ [ y for y in list(x) ] for x in input_words ]
        self.weights = weights

    def compute_score(self, candidate):
        similarity_scores = self.compute_similarity_scores(candidate)
        weighted_sum = self.calculate_weighted_sum(similarity_scores)
        return weighted_sum, similarity_scores

    def compute_similarity_scores(self, candidate):
      chars = [ x for x in list(candidate) ]
      return [ self.compute_similarity_score(candidate, chars, word_chars) \
               for word_chars in self.input_words_chars ]

    def compute_similarity_score(self, candidate, candidate_chars, input_chars):
        lcs_len = lcs_length(candidate_chars, input_chars)
        if lcs_len < 2:
            score = 0
        elif lcs_len == 2:
            score = self.score_dyad_by_pattern(candidate, ''.join(input_chars))
        else:
            score = lcs_len
        return float(score) / len(input_chars)

    def score_dyad_by_pattern(self, candidate, input_word):
        patterns = self.gismu_dyad_patterns(candidate)
        return 2 if self.matches_any_pattern(patterns, input_word) else 0

    def gismu_dyad_patterns(_, gismu):
        patterns = []
        for i, c in enumerate(gismu[:-2]):
            patterns.append('%s(%s|.%s)' % (c, gismu[i + 1], gismu[i + 2]))
        patterns.append('%s%s' % (gismu[-2], gismu[-1]))
        return patterns

    def matches_any_pattern(_, patterns, word):
        return next((True for pat in patterns if re.search(pat, word)), False)

    def calculate_weighted_sum(self, scores):
        # Multiply each score by the given weight, then sum weighted scores
        return reduce(float.__add__, map(float.__mul__, scores, self.weights))

class GismuMatcher:

    def __init__(self, gismus, stem_length = 4):
        self.gismus = gismus
        self.stem_length = stem_length

    def find_similar_gismu(self, candidate):
        candidate = candidate.rstrip()
        patterns = map(lambda x:SIMILARITIES.get(x, '.'), candidate)
        # e.g. "rekpa" =~ /l.[gx][bf]./ where . matches NOTHING

        gismu = None
        found_match = False
        for gismu in self.gismus:
            found_match = self.match_gismu(gismu, candidate, patterns)
            if found_match:
                break
        return gismu if found_match else None

    def match_gismu(self, gismu, candidate, structural_patterns):
        return self.match_stem(gismu, candidate) or \
          self.match_structure(gismu, candidate, structural_patterns)

    def match_stem(self, gismu, candidate):
        return candidate[:self.stem_length] == gismu[:self.stem_length]

    # Apply similarity pattern to each position in gismu if others letters equal
    #
    # /PAT/ekpa
    # r/PAT/kpa
    # re/PAT/pa
    # rek/PAT/a
    # rekp/PAT/
    #
    def match_structure(self, gismu, candidate, structural_patterns):
        similar = False
        common_len = min(len(candidate), len(gismu))
        for i in xrange(common_len):
            if self.strings_match_except(gismu, candidate, i, common_len) and \
              self.match_structural_pattern(gismu[i], structural_patterns[i]):
                similar = True
                break
        return similar

    def strings_match_except(_, x, y, i, j):
      return x[:i] == y[:i] and x[i+1:j] == y[i+1:j]

    def match_structural_pattern(_, letter, pattern):
        if pattern == '.':
            return False
        return letter in pattern



gismu_score.py
#!/usr/bin/env python

# Lojban gismu candidate creation and evaluation script
# Version 0.5

# Copyright 2014 Riley Martinez-Lynch, except where
# Copyright 2012 Arnt Richard Johansen.
# Distributed under the terms of the GPL v3.

# Usage:
#
#   python gismu_score.py uan rakan ekspekt esper predpologa mulud
#

import sys
import re
import threading
import Queue

from optparse import OptionParser

from gismu_utils import C, V, LANGUAGE_WEIGHTS, \
  GismuGenerator, GismuScorer, GismuMatcher

from marshal import dump

VERSION = "v0.5"

DEFAULT_LANGUAGE_WEIGHTS = LANGUAGE_WEIGHTS['1995'] # as they appear in CLL

def main(words, params):

    if params.all_letters:
        (c, v) = (C, V)
    else:
        (c, v) = letters_for_words(words)

    # "In practice, we only scored forms that used the letters/phonemes in the
    #  source languages, which greatly reduced the computation."
    # (http://mail.lojban.org/lists/lojban-list/msg26536.html)

    shapes = re.split('\s*,\s*', params.shapes)
    candidate_iterator = GismuGenerator(c, v, shapes).iterator()

    print >>sys.stderr, "Generating candidates..."
    # pre-extracting candidates from the iterator is not needed downstream
    # but lets us inform the user how many candidates will be evaluated
    candidates = list(candidate_iterator)
    print >>sys.stderr, "%d candidates generated." % len(candidates)

    weights = [float(weight) for weight in re.split("\s*,\s*", params.weights)]

    scorer = GismuScorer(words, weights)
    if params.workers == 1:
        print >>sys.stderr, "Scoring candidates..."
        scores = compute_scores(candidates, scorer, params)
    else:
        print >>sys.stderr, \
          "Scoring candidates with %d workers..." % params.workers
        scores = compute_scores_threaded(candidates, scorer, params)

    print >>sys.stderr, "Sorting scores..."
    scores.sort(lambda x,y:cmp(y[0], x[0]))

    print >>sys.stderr, ""
    print >>sys.stderr, "10 first gismu candidates are:"
    print >>sys.stderr, ""
    for record in scores[:10]:
        print >>sys.stderr, record

    if params.gismu_list_path:
        print >>sys.stderr, "Reading list of gismu... "
        gismus = [line.strip() for line in file(params.gismu_list_path)]
        matcher = GismuMatcher(gismus)
        print >>sys.stderr, "Exluding candidates similar to existing gismu..."
        candidate = deduplicate_candidates(matcher, scores)
        if candidate == None:
            print >>sys.stderr, "No suitable candidates found."
        else:
            print >>sys.stderr, "The winner is....\n"
            print candidate.upper()

    if params.output_path:
        print >>sys.stderr, "\rDumping scores to %s" % params.output_path
        sink = open(params.output_path, 'w')
        dump(scores, sink)
        sink.close()

def letters_for_words(words):
    word_set = set([ l for word in words for l in list(word) ])
    return word_set.intersection(C), word_set.intersection(V)

def compute_scores(candidates, scorer, params):
    quiet = params.quiet
    scores = []
    for i, candidate in enumerate(candidates):
        score = compute_score(scorer, candidate)
        scores.append(score)
        if (not quiet) and (i % 100 == 0):
            print >>sys.stderr, "\010" * 9, # backspace
            print >>sys.stderr, i + 1,
    return scores

def compute_score(scorer, candidate):
    weighted_sum, language_scores = scorer.compute_score(candidate)
    return (weighted_sum, candidate, language_scores)

def compute_scores_threaded(candidates, scorer, params):
    input_queue = Queue.Queue()
    for candidate in candidates:
        input_queue.put(candidate)
    output_queue = Queue.Queue()

    workers = []
    for x in xrange(params.workers):
        worker = QueueWorker(input_queue, output_queue, scorer, params)
        worker.start()
        workers.append(worker)
    for worker in workers:
        worker.join()
    return read_scores_from_queue(output_queue)

def read_scores_from_queue(score_queue):
    scores = []
    queue_empty = False
    while not queue_empty:
        try:
            score = score_queue.get_nowait()
            scores.append(score)
        except Queue.Empty:
            queue_empty = True
    return scores

def deduplicate_candidates(matcher, scores):
    unique_candidate = None
    for (score, candidate, _) in scores:
        gismu = matcher.find_similar_gismu(candidate)
        if gismu == None:
            unique_candidate = candidate
            break
        else:
            print >>sys.stderr, \
              "Candidate '%s' too much like gismu '%s'." % (candidate, gismu)
    return unique_candidate

##

class QueueWorker(threading.Thread):

    def __init__(self, input_queue, output_queue, scorer, params):
        self.input_queue = input_queue
        self.output_queue = output_queue
        self.scorer = scorer
        self.params = params
        threading.Thread.__init__(self)

    def run(self):
        quiet = self.params.quiet
        i = 0
        while 1:
            candidate = self.read_input()
            if candidate == None:
                break
            self.process_input(candidate)

            i += 1
            if (not quiet) and (i % 100 == 0):
                print >>sys.stderr, "\010" * 9, # backspace
                print >>sys.stderr, i,

    def read_input(self):
        candidate = None
        try:
            candidate = self.input_queue.get(True, 2) # block for up to 2s
        except Queue.Empty:
            pass
        return candidate

    def process_input(self, candidate):
        weighted_sum, language_scores = self.scorer.compute_score(candidate)
        self.output_queue.put((weighted_sum, candidate, language_scores))

##

def check_weights_option(option, opt_str, value, parser):
    if re.match('(\d{4}|finprims)$', value):
        if value in LANGUAGE_WEIGHTS:
            print >>sys.stderr, "Using language weights from %s..." % value
            value = ",".join([ str(x) for x in LANGUAGE_WEIGHTS[value] ])
        else:
            raise ValueError("No weights registered for %d" % year)
    else:
        weights = re.split('\s*,\s*', value)
        if len(weights) < 2:
            raise ValueError("%s must include at least 2 values" % opt_str)
        [ check_weight_option(opt_str, weight) for weight in weights ]
    setattr(parser.values, option.dest, value)

def check_weight_option(opt_str, weight_string):
    error = False
    try:
        weight = float(weight_string)
        error = weight <= 0
    except:
        error = True
    if error:
        raise ValueError("Values for %s must be numbers greater than zero" % opt_str)

def check_shapes_option(option, opt_str, value, parser):
    shapes = re.split('\s*,\s*', value)
    for shape in shapes:
        if len(shape) < 4:
            raise ValueError("Value for %s must contain at least 4 letters" % opt_str)
        if not re.match('[cv]+$', shape, re.I):
            raise ValueError("Value for %s must consist only of 'c' and 'v'" % opt_str)
    setattr(parser.values, option.dest, value)

def validate_words(words, params):
    weight_count = len(re.split("\s*,\s*", params.weights))
    if len(words) != weight_count:
        raise ValueError("Expected %d words as input" % weight_count)
    for word in words:
        if len(word) < 2:
            raise ValueError("Input words must be at least two letters long")

if __name__ == '__main__':

    usage_fmt = "usage: %prog [ options ] { input words }"
    options = OptionParser(usage=usage_fmt, version="%prog " + VERSION)

    options.add_option("-q", "--quiet",
                       action="store_true", dest="quiet")

    options.add_option("-a", "--all-letters",
                       action="store_true", dest="all_letters")

    default_weights = ",".join([ str(x) for x in DEFAULT_LANGUAGE_WEIGHTS ])
    options.add_option("-w", "--weights",
                       default=default_weights, dest="weights",
                       type="string", action="callback",
                       callback=check_weights_option)

    options.add_option("-s", "--shapes",
                       default="ccvcv,cvccv", dest="shapes",
                       type="string", action="callback",
                       callback=check_shapes_option)

    options.add_option("-n", "--number-workers",
                       type="int", default="1", dest="workers")

    options.add_option("-d", "--deduplicate",
                       type="string", dest="gismu_list_path")

    options.add_option("-o", "--output",
                       type="string", dest="output_path")

    error = False

    try:
        (params, words) = options.parse_args()
        validate_words(words, params)
    except Exception, e: # NOTE: python 2.5 syntax for jython (sorry, python3!)
        print >>sys.stderr, "\n%s\n" % e
        error = True

    if error:
        options.print_help()
        print >>sys.stderr
        exit()

    main(words, params)

